ttnn/cpp/ttnn/operations/moreh/moreh_adamw/device/moreh_adamw_device_operation.cpp

```diff
@@ -0,0 +1,134 @@
+        tt::operations::primary::check_tensor(
+            tensor_args.max_exp_avg_sq_out.value(), "moreh_adamw", "max_exp_avg_sq_out");
+    }
+}
+
+void MorehAdamWDeviceOperation::validate_on_program_cache_hit(
+    const operation_attributes_t& attributes, const tensor_args_t& tensor_args) {}
+
+MorehAdamWDeviceOperation::shape_return_value_t MorehAdamWDeviceOperation::compute_output_shapes(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    auto input_tensor_shape = tensor_args.param_in.get_shape();
+
+    return {
+        input_tensor_shape,
+        input_tensor_shape,
+        input_tensor_shape,
+        input_tensor_shape,
+    };
+}
+
+MorehAdamWDeviceOperation::tensor_return_value_t MorehAdamWDeviceOperation::create_output_tensors(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    auto output_shapes = compute_output_shapes(operation_attributes, tensor_args);
+    auto dtype = tensor_args.param_in.get_dtype();
+    Layout layout{Layout::TILE};
+    auto device = tensor_args.param_in.device();
+
+    tensor_return_value_t result;
+
+    auto memory_config = operation_attributes.memory_config;
+
+    if (tensor_args.param_out.has_value()) {
+        result.push_back(tensor_args.param_out.value());
+    } else {
+        result.push_back(create_device_tensor(output_shapes.at(0), dtype, layout, device, memory_config));
+    }
+
+    if (tensor_args.exp_avg_out.has_value()) {
+        result.push_back(tensor_args.exp_avg_out.value());
+    } else {
+        result.push_back(create_device_tensor(output_shapes.at(1), dtype, layout, device, memory_config));
+    }
+
+    if (tensor_args.exp_avg_sq_out.has_value()) {
+        result.push_back(tensor_args.exp_avg_sq_out.value());
+    } else {
+        result.push_back(create_device_tensor(output_shapes.at(2), dtype, layout, device, memory_config));
+    }
+
+    if (tensor_args.max_exp_avg_sq_out.has_value()) {
+        result.push_back(tensor_args.max_exp_avg_sq_out.value());
+    } else if (operation_attributes.amsgrad) {
+        result.push_back(create_device_tensor(output_shapes.at(3), dtype, layout, device, memory_config));
+    }
+
+    return std::move(result);
+}
+
+std::tuple<MorehAdamWDeviceOperation::operation_attributes_t, MorehAdamWDeviceOperation::tensor_args_t>
+MorehAdamWDeviceOperation::invoke(
+    const Tensor& param_in,
+    const Tensor& grad,
+    const Tensor& exp_avg_in,
+    const Tensor& exp_avg_sq_in,
+
+    float lr,
+    float beta1,
+    float beta2,
+    float eps,
+    float weight_decay,
+    uint32_t step,
+    bool amsgrad,
+
+    const std::optional<const Tensor> max_exp_avg_sq_in,
+    const std::optional<const Tensor> param_out,
+    const std::optional<const Tensor> exp_avg_out,
+    const std::optional<const Tensor> exp_avg_sq_out,
+    const std::optional<const Tensor> max_exp_avg_sq_out,
+    const MemoryConfig& memory_config,
```

Comment:
```suggestion
    const std::optional<MemoryConfig>& memory_config,
```
I asked about `invoke` and received a response.
"We prefer `optional<MemoryConfig>` argument. When nullopt, we default to the memory config of the first argument.
There are still some operations which do not follow the guidance."

Reply:
I see it now, updated it.